compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -4435,8 +4438,11 @@ private BLangBlockStmt desugarForeachToWhile(BLangForeach foreach, BLangSimpleVa
         // However, we are within the while loop. hence the $result$ can never be nil nor error.
         // Therefore cast $result$ to non-nilable type. i.e `int item = <>$result$.value;`
         BLangFieldBasedAccess valueAccessExpr = getValueAccessExpression(foreach.pos, foreach.varType, resultSymbol);
-        valueAccessExpr.expr = addConversionExprIfRequired(valueAccessExpr.expr,
-                types.getSafeType(valueAccessExpr.expr.type, true, false));
+
+        BLangExpression expr = valueAccessExpr.expr;
+        // Since `$result$` is always a record (with a single `value` field), add a cast to `map<any|error>`
+        // to avoid casting to an anonymous type - https://github.com/ballerina-platform/ballerina-lang/issues/28262.
+        valueAccessExpr.expr = addConversionExprIfRequired(expr, symTable.mapAllType);
```

Comment:
Out of curiosity, what happens if we don't have this conversion here? Issues in jvm codegen or does it get re-added somehow via desugaring the desugared constructs?

Reply:
So the type of `expr` (`$result$`) here is the union of a record and nil - `record {|T value;|}?`. Although semantically it is invalid to do `$result$.value` in such a scenario, desugaring proceeds as a field access on a record union due to the check at https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L5171. And then at org.wso2.ballerinalang.compiler.bir.codegen.JvmInstructionGen#generateMapLoadIns we don't seem to be adding a cast so probably works as is there?